File: NeetCode/Binary_Search/Search_In_Rotated_Array_II/Search_In_Rotated_Array_II.cpp
// Search_In_Rotated_Array_II.cpp
//
// Compile locally with:
// clang++ Search_In_Rotated_Array_II.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool search(vector<int> &nums, int target)
    {
        int left(0), mid(0);
        int right = nums.size() - 1;

        while (left <= right)
        {
            mid = left + (right - left) / 2;

            if (nums[mid] == target || nums[left] == target || nums[right] == target)
            {
                return true;
            }

            if (nums[mid] == nums[left] && nums[mid] == nums[right])
            {
                left++;
                right--;
            }
            else if (nums[left] <= nums[mid])
            {
                // sorted left half
                if (nums[left] < target && target < nums[mid])
                {
                    right = mid - 1;
                }
                else
                {
                    left = mid + 1;
                }
            }
            else
            {
                if (nums[mid] < target && target < nums[right])
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }
        }
        return false;
    }
};

int main()
{
    cout << "Running Search_In_Rotated_Array_II.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: NeetCode/Binary_Search/Search_In_Rotated_Array_II/search__in__rotated__array__i_i.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = (left + right) // 2

            if nums[mid] == target or nums[left] == target or nums[right] == target:
                return True

            if nums[left] == nums[mid] == nums[right]:
                left += 1
                right -= 1

            elif nums[left] < nums[mid]:
                # left half is sorted
                if nums[left] < target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1

            else:
                # right half sorted
                if nums[mid] < target < nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return False


if __name__ == "__main__":
    sol = Solution()
    nums = [1, 0, 1, 1, 1]
    target = 0
    print(sol.search(nums, target))
    print("Running Solution...")
